### PR TITLE
[REEF-1133] Wake Time incorrectly implements Compare

### PR DESCRIPTION
**Pull request by afchung**
This addressed the issue by
  * Implementing GetHashCode and CompareTo only based on long TimeStamp.
  * Adding comment in Equals regarding comparing subclasses of Time.
  * Adding a PriorityQueue implementation.
  * Change RuntimeClock to use PriorityQueue instead of SortedSet.
  * Adding tests for PriorityQueue and Time.

JIRA:
  [REEF-1133](https://issues.apache.org/jira/browse/REEF-1133)

### CODE REVIEW

**Comment by markusweimer** [comment]
I'll have a look.

File: lang/cs/Org.Apache.REEF.Utilities/Collections/PriorityQueue.cs [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+using System.Collections;
+using System.Collections.Generic;
+
+namespace Org.Apache.REEF.Utilities.Collections
+{
+    /// <summary>
+    /// A simple priority queue implementation, where the head of the queue is the 
+    /// smallest element in the priority queue.
+    /// </summary>
+    public sealed class PriorityQueue<T> : ICollection<T> where T : IComparable<T>
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Utilities/Collections/PriorityQueue.cs]: How is this related to the issue being fixed here?

File: lang/cs/Org.Apache.REEF.Wake.Tests/TimeTest.cs [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        [Fact]
+        public void BadTime()
+        {
+            Assert.Throws<ArgumentException>(() => new StartTime(-1));
+        }
+
+        [Fact]
+        public void SimpleTimeComparison()
+        {
+            long ts = new Random().Next(0, 100);
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Wake.Tests/TimeTest.cs]: Please add a seed to the random, to make the tests repeatable.

File: lang/cs/Org.Apache.REEF.Wake/Time/Time.cs [hunk 3]
```diff
@@ -29,6 +29,11 @@ public abstract class Time : IComparable<Time>
     {
         public Time(long timeStamp)
         {
+            if (timeStamp < 0)
+            {
+                throw new ArgumentException("Time must have timeStamp of at least 0.");
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Wake/Time/Time.cs]: If that is the case, shouldn't we use unsigned types to begin with?

File: lang/cs/Org.Apache.REEF.Wake/Time/Time.cs [hunk 4]
```diff
@@ -50,33 +55,21 @@ public override bool Equals(object obj)
             {
                 return true;
             }
-            Time other = obj as Time;
+            
+            // Note: This allows for the very strange semantic where
+            // two different subclasses might return true for Equals.
+            var other = obj as Time;
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Wake/Time/Time.cs]: I don't think this is right. Instances of different subclasses of `Time` should be unequal in general, unless they specifically are coded to be compared against one another.

Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Wake/Time/Time.cs]: Also, `Time` really should be an interface. That way, we don't ever run into this issue. Not all implementations of `Time` will be using `TimeStamp` to store the time.

**Comment by markusweimer** [comment]
The changes seem fine to me, I left some minor comments.

I am quite confused about the use of `Time` in Wake.NET. I will start a thread on the dev list for that.

File: lang/cs/Org.Apache.REEF.Utilities/Collections/PriorityQueue.cs (hunk 1 above)
Comment (afchung) [inline lang/cs/Org.Apache.REEF.Utilities/Collections/PriorityQueue.cs]: The `GetHashCode` method in `Time` because it's a bizarre workaround to allow multiple instances of the same object in `SortedSet`. Using a `PriorityQueue` allows us to properly impose order on the `Time` objects.

File: lang/cs/Org.Apache.REEF.Wake/Time/Time.cs (hunk 4 above)
Comment (afchung) [inline lang/cs/Org.Apache.REEF.Wake/Time/Time.cs]: I agree with you here, but it seems that different types of `Time` need to be compared with each other in the `RuntimeClock`, hence the original workaround.

Comment (afchung) [inline lang/cs/Org.Apache.REEF.Wake/Time/Time.cs]: Perhaps we should have an `ITime` interface and an `IComparer<ITime>` class instead?

File: lang/cs/Org.Apache.REEF.Utilities/Collections/PriorityQueue.cs (hunk 1 above)
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Utilities/Collections/PriorityQueue.cs]: Ah, yes, that makes sense.

File: lang/cs/Org.Apache.REEF.Wake/Time/Time.cs (hunk 4 above)
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Wake/Time/Time.cs]: Let's do this as a separate discussion. The change as is is minimally invasive and fixes a bug, so we should proceed.

**Comment by markusweimer** [comment]
@afchung Let me know when this is ready for another round.

File: lang/cs/Org.Apache.REEF.Wake/Time/Time.cs (hunk 4 above)
Comment (afchung) [inline lang/cs/Org.Apache.REEF.Wake/Time/Time.cs]: OK, sounds good, I'll address the comments in this PR first then!

File: lang/cs/Org.Apache.REEF.Wake/Time/Time.cs (hunk 3 above)
Comment (afchung) [inline lang/cs/Org.Apache.REEF.Wake/Time/Time.cs]: I think we should just stick with `long` right now to avoid breaking the API. If we end up implementing `ITime`, we can then use `ulong`.

**Comment by afchung** [comment]
@markusweimer I've addressed your comments, please have another look. Thanks!

**Comment by markusweimer** [comment]
Looks good, I will test & merge.
